### PR TITLE
Don't trim ShouldSerializeXXX and ResetXXX members that may be called with reflection

### LINKED ISSUE #102244: Do not trim private methods used by the designer
### Description

Copied from https://github.com/dotnet/winforms/issues/11314

Private methods whose names follow the pattern `ShouldSerialize<PropertyName>` or `Reset<ProppertyName>` should not be trimmed by the linker because they are used by the TypeDescriptor via reflection. Additional conditions: Winforms designer is interested in types derived from `IComponent`. Property should not be hidden from serializers via SerializerVisibility attribute. Property should be public.

TypeDescriptor accesses these methods [here](https://source.dot.net/#System.ComponentModel.TypeConverter/System/ComponentModel/MemberDescriptor.cs,262359213ad5b14e)

Suggested fix from @ericstj 

> The fix is to add a file here [runtime/src/libraries/System.Data.Common/src/ILLink at bad00cf23ec49a2607776ffd6e1810c4dbf540b3 · dotnet/runtime (github.com)](https://github.com/dotnet/runtime/tree/bad00cf23ec49a2607776ffd6e1810c4dbf540b3/src/libraries/System.Data.Common/src/ILLink)
 
> That looks similar to [runtime/src/libraries/System.ComponentModel.TypeConverter/src/ILLink/ILLink.Descriptors.LibraryBuild.xml at bad00cf23ec49a2607776ffd6e1810c4dbf540b3 · dotnet/runtime (github.com)](https://github.com/dotnet/runtime/blob/bad00cf23ec49a2607776ffd6e1810c4dbf540b3/src/libraries/System.ComponentModel.TypeConverter/src/ILLink/ILLink.Descriptors.LibraryBuild.xml)

> Search pattern for [ShouldSerialize](https://github.com/search?q=repo%3Adotnet%2Fruntime%20%2Fprivate%20bool%20ShouldSerialize%5B%5E%5C(%20%5D%2B%5C(%5C)%2F&type=code )

> Search pattern for [Reset](https://github.com/search?q=repo%3Adotnet%2Fruntime+%2Fprivate+void+Reset%5B%5E%5C%28+%5D%2B%5C%28%5B%5E%2C%5D%2B%5C%29%2F&type=code)

Searches yeld some false positives if the class is not designable, not an `IComponent`, or class does not have the <Name> public, serializable property, or method is actually called directly.  For each property we need to examine metadata in the trimmed assembly to see if the methods were removed.

### Reproduction Steps

This works on 4.8.1 but doesn't on net6

```c#
using System;
using System.ComponentModel;
using System.Data;

namespace ConsoleApp1
{
    internal class Program
    {
        static void Main(string[] args)
        {
            DataColumn dataColumn1 = new DataColumn
            {
                ColumnName = "dataColumn1",
                DataType = typeof(DateTime)
            };

            var properties = TypeDescriptor.GetProperties(dataColumn1);
            var property = properties[nameof(DataColumn.DefaultValue)];
            if (property != null)
            {
                bool shouldSerialize = property.ShouldSerializeValue(dataColumn1);
                Console.WriteLine($"ShouldSerialize default? expected {false} actual {shouldSerialize}");

                dataColumn1.DefaultValue = DateTime.MinValue;
                shouldSerialize = property.ShouldSerializeValue(dataColumn1);
                Console.WriteLine($"ShouldSerialize changed? expected {true} actual {shouldSerialize}");

                // Reset method is not available
                property.ResetValue(dataColumn1);
                Console.WriteLine($"Reset? expected {DateTime.MinValue} actual {dataColumn1.DefaultValue}");
            }

            property = properties[nameof(DataColumn.Caption)];
            if (property != null)
            {
                bool shouldSerialize = property.ShouldSerializeValue(dataColumn1);
                Console.WriteLine($"ShouldSerialize default? expected {false} actual {shouldSerialize}");

                dataColumn1.Caption = "Caption";
                shouldSerialize = property.ShouldSerializeValue(dataColumn1);
                Console.WriteLine($"ShouldSerialize changed? expected {true} actual {shouldSerialize}");

                // Reset method is available
                property.ResetValue(dataColumn1);
                Console.WriteLine($"Reset? expected {nameof(dataColumn1)} actual {dataColumn1.Caption}");
            }

            property = properties[nameof(DataColumn.Namespace)];
            if (property != null)
            {
                bool shouldSerialize = property.ShouldSerializeValue(dataColumn1);
                Console.WriteLine($"ShouldSerialize default? expected {false} actual {shouldSerialize}");

                dataColumn1.Namespace = "Namespace";
                shouldSerialize = property.ShouldSerializeValue(dataColumn1);
                Console.WriteLine($"ShouldSerialize changed? expected {true} actual {shouldSerialize}");

                // Reset method is available
                property.ResetValue(dataColumn1);
                Console.WriteLine($"Reset? expected '' actual `{dataColumn1.Namespace}`");
            }

            _ = Console.ReadLine();
        }
    }
}
```

### Expected behavior

ShouldSerialize, Reset methods are present in the assembly

### Actual behavior

ShouldSerialize, Reset methods are trimmed 

### Regression?

The end scenario regressed between the InProc and OOP designers.

### Known Workarounds

none

### Configuration

WIndows

### Other information

Trimmed ShouldSerialize causes serialization using BinaryFormatter. The fix should be serviced.

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/runtime/issues/102244

This change makes sure these members exist in the inbox version of the `System.Data.Common` assembly.

The `TypeDescriptor` reflection-based model, the XML serializer, and the Json.NET serializer use these naming conventions to call the members with reflection.

### CODE REVIEW

**Review by MichalStrehovsky — CHANGES_REQUESTED** [review]
This doesn't look like the right fix. https://github.com/dotnet/runtime/issues/102244#issuecomment-2138184386 and the subsequent discussion discuss why.

I think this should use ILLink.LibraryBuild file like the original issue was suggesting so that these get preserved when we build the repo. After repo build it can be trimmed like anything else. ILLink.LibraryBuild is designed for these cases.

These DynamicDependency will keep things unconditionally in trimmed apps even if nobody uses TypeDescriptor/Json.NET/etc. It's a global fix for a local problem. It will regress the size for everyone, including the well behaved apps that don't use trim unsafe code this tries to fix. Everything that could need this should already be marked as trim unsafe. We don't need trim unsafe code to work in trimmed apps, it is okay and expected that trim unsafe code will be broken after trimming. It is the responsibility of whoever is turning on trimming on trim unsafe code to author XML descriptors if they wish to take the hard path.

> The `TypeDescriptor` reflection-based model and the XML and Json.NET serializers use these conventions.

These are also trim unsafe and not expected to work after trimming. The issue this PR tries to address is a drop in the bucket, there are more fundamental trimming problems with these serializers.

**Comment by steveharter** [comment]
> This doesn't look like the right fix. https://github.com/dotnet/runtime/issues/102244#issuecomment-2138184386 and the subsequent discussion discuss why.This doesn't look like the right fix. https://github.com/dotnet/runtime/issues/102244#issuecomment-2138184386 and the subsequent discussion discuss why.

I assume the change requested here is to move the DynamicDependency from source to XML like this:
```xml
    <type fullname="System.Data.DataSet">
      <method signature="System.Void .ctor()">
        <attribute fullname="System.Diagnostics.CodeAnalysis.DynamicDependencyAttribute">
          <argument>ResetRelations</argument>
        </attribute>
```

however, when I do that it appears trimming already occurred, meaning in the example above I'll get this error when building:
```
C:\git\ComponentModel\src\libraries\System.Data.Common\src\System\Data\DataSet.cs(47,9): error IL2037: System.Data.DataSet.DataSet(): No members were resolved for 'ResetRelations' on type 'System.Data.DataSet'. [C:\git\ComponentModel\src\libraries\sfx.proj]
```

**Comment by jkotas** [comment]
> I assume the change requested here is to move the DynamicDependency from source to XML like this:

Yes, it should be in `ILLink.Descriptors.LibraryBuild.xml`.

> however, when I do that it appears trimming already occurred, meaning in the example above I'll get this error when building:

I am not sure where this error is coming from. Could you please apply the change so that we can see it?

**Comment by steveharter** [comment]
> Yes, it should be in ILLink.Descriptors.LibraryBuild.xml.

I saw that right after I posted; I was using the wrong xml file. I added `src\libraries\System.Data.Common\src\ILLink` which is working.

File: src/libraries/System.Data.Common/src/ILLink/ILLink.Descriptors.LibraryBuild.xml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
+<linker>
+  <assembly fullname="System.Data.Common, PublicKeyToken=b03f5f7f11d50a3a">
```
Comment (MichalStrehovsky) [inline src/libraries/System.Data.Common/src/ILLink/ILLink.Descriptors.LibraryBuild.xml:3]: Nit: we don't put the public key token in any of our descriptor files.

```suggestion
  <assembly fullname="System.Data.Common">
```

**Review by MichalStrehovsky — APPROVED** [review]
I'm not reviewing whether this list is complete, but the approach looks good to me! Thanks!

**Comment by steveharter** [comment]
This should be considered for servicing one verified.

**Comment by SamMonoRT** [comment]
@roji @AndriySvyryd fyi --

**Pull request merged**